core/net/ipv6/websocket.c

```diff
@@ -0,0 +1,719 @@
+        /* Need to keep parsing the incoming data to check for more
+           frames, if the incoming datalen is > than len. */
+        if(datalen > 0) {
+          PRINTF("XXX 2 again (datalen %d s->left %d)\n", datalen, (int)s->left);
+          websocket_http_client_datahandler(client_state,
+                                            data, datalen);
+
+        }
+      }
+    }
+  }
+}
+/*---------------------------------------------------------------------------*/
+static void
+init(void)
+{
+  static uint8_t inited = 0;
+  if(!inited) {
+    process_start(&websocket_process, NULL);
+    list_init(websocketlist);
+    inited = 1;
+  }
+}
+/*---------------------------------------------------------------------------*/
+void
+websocket_init(struct websocket *s)
+{
+  init();
+  websocket_http_client_init(&s->s);
+}
+/*---------------------------------------------------------------------------*/
+void
+websocket_set_proxy(struct websocket *s,
+                    const uip_ipaddr_t *addr, uint16_t port)
+{
+  websocket_http_client_set_proxy(&s->s, addr, port);
+}
+/*---------------------------------------------------------------------------*/
+websocket_result_t
+websocket_open(struct websocket *s, const char *url,
+               const char *subprotocol, const char *hdr,
+	       websocket_callback c)
+{
+  int ret;
+  char host[MAX_HOSTLEN];
+  char path[MAX_PATHLEN];
+  uint16_t port;
+  uip_ipaddr_t addr;
+
+  init();
+
+  if(s == NULL) {
+    return WEBSOCKET_ERR;
+  }
+
+  if(s->state != WEBSOCKET_STATE_CLOSED) {
+    PRINTF("websocket_open: closing websocket before opening it again.\n");
+    websocket_close(s);
+  }
+  s->callback = c;
+
+  if(parse_url(url, host, &port, path)) {
+    list_add(websocketlist, s);
+    websocket_http_client_register(&s->s, host, port, path, subprotocol, hdr);
+
+    /* First check if the host is an IP address. */
+    if(uiplib_ip4addrconv(host, (uip_ip4addr_t *)&addr) == 0 &&
+       uiplib_ip6addrconv(host, (uip_ip6addr_t *)&addr) == 0) {
+      /* Try to lookup the hostname. If it fails, we initiate a hostname
+	 lookup and print out an informative message on the
+	 statusbar. */
+      ret = resolv_lookup(host, NULL);
+      if(ret != RESOLV_STATUS_CACHED) {
+	resolv_query(host);
+	s->state = WEBSOCKET_STATE_DNS_REQUEST_SENT;
+	PRINTF("Resolving host...\n");
+	return WEBSOCKET_OK;
+      }
+    }
+
+    PROCESS_CONTEXT_BEGIN(&websocket_process);
+    ret = start_get(s);
+    PROCESS_CONTEXT_END();
+    return ret;
+  }
+  return -1;
+}
+/*---------------------------------------------------------------------------*/
+void
+websocket_close(struct websocket *s)
+{
+  websocket_http_client_close(&s->s);
+  s->state = WEBSOCKET_STATE_CLOSED;
+}
+/*---------------------------------------------------------------------------*/
+static int
+send_data(struct websocket *s, const void *data,
+          uint16_t datalen, uint8_t data_type_opcode)
+{
+  uint8_t buf[WEBSOCKET_MAX_MSGLEN + 4 + 4];
```

Comment:
Where does the `buf` lenght comes from? perhaps adding a comment here explaining the `4 +4`